From 27bd8ffbd5a421d3f76a27f4a8de54c57cbbb373 Mon Sep 17 00:00:00 2001
From: Tian Yang <tiany@codeaurora.org>
Date: Mon, 28 Sep 2015 17:23:31 -0500
Subject: [PATCH] skbuff_debug: dump ftrace data on error

If ftrace is enabled, dump the call trace as it can assist
with debugging

Change-Id: I0f77faa5f67d56461e98ba673776a2cb7d7e67cd
Signed-off-by: Matthew McClintock <mmcclint@codeaurora.org>
Signed-off-by: Casey Chen <kexinc@codeaurora.org>
---
 net/core/skbuff_debug.c | 8 ++++++--
 net/core/skbuff_debug.h | 2 +-
 2 files changed, 7 insertions(+), 3 deletions(-)

--- a/net/core/skbuff_debug.c
+++ b/net/core/skbuff_debug.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2013-2014, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2015, The Linux Foundation. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for any
  * purpose with or without fee is hereby granted, provided that the above
@@ -26,6 +26,7 @@ static bool skbuff_debugobj_fixup(void *
 static int skbuff_debugobj_fixup(void *addr, enum debug_obj_state state)
 #endif
 {
+	ftrace_dump(DUMP_ALL);
 	WARN(1, "skbuff_debug: state = %d, skb = 0x%p\n", state, addr);
 #ifdef CONFIG_ARM64
 	return true;
@@ -46,9 +47,11 @@ inline void skbuff_debugobj_activate(str
 {
 	int ret = debug_object_activate(skb, &skbuff_debug_descr);
 
-	if (ret)
+	if (ret) {
+		ftrace_dump(DUMP_ALL);
 		WARN(1, "skb_debug: failed to activate err = %d skb = 0x%p\n",
 		     ret, skb);
+	}
 }
 
 inline void skbuff_debugobj_init_and_activate(struct sk_buff *skb)
@@ -66,6 +69,7 @@ inline void skbuff_debugobj_deactivate(s
 		return;
 	}
 
+	ftrace_dump(DUMP_ALL);
 	WARN(1, "skbuff_debug: deactivating inactive object skb 0x%p state=%d\n",
 	     skb, obj_state);
 }
--- a/net/core/skbuff_debug.h
+++ b/net/core/skbuff_debug.h
@@ -1,5 +1,5 @@
 /* 
- * Copyright (c) 2013-2014, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2015, The Linux Foundation. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for any
  * purpose with or without fee is hereby granted, provided that the above
